media-libs/openimageio: version bumps to 1.6.18 and 1.7.17

dracwyrm opened the pull request:
**profiles/base/package.use.mask: Change media-libs/openimageio mask**
Newer versions have a fix for python detection.

**media-libs/openimageio: version bumps to 1.6.18 and 1.7.17**
* Version 1.6.18 is the last of the RB-1.6 series, but add for
  compatibility with all existing programs.
* v1.6.18 only: Add patch to support Boost 1.65 changes
* The follow applies to both versions:
*   Patch to use GNUInstallDirs
*   Patch to fix boost and python detection on Gentoo
*   Added PTex and Field3D support
*   Added python module building support in ebuild
*   Added SIMD optimization (cpu flags) selection in ebuild

Closes: https://bugs.gentoo.org/596268

**Additional notes not in commit message:**
I'm not not completely sure, but I think the 1.6.18 ebuild fixes the following bug reports:
https://bugs.gentoo.org/show_bug.cgi?id=627302
https://bugs.gentoo.org/show_bug.cgi?id=616086
https://bugs.gentoo.org/show_bug.cgi?id=616090
https://bugs.gentoo.org/show_bug.cgi?id=587332
These will need to be double checked. I'm able to compile 1.6.18 with Boost 1.65 and other modern libraries.

dracwyrm commented:
If there are no issues with version 1.6.18, then it should be made stable as quick as possible to replace 1.6.13. It should be compatible with everything that 1.6.13 is. I cannot guarantee that 1.7.17 is, which is why I fixed 1.6.18 with all the backported patches.

dracwyrm commented:
Pinging @SoapGentoo

devurandom commented:
In my overlay, I have an ebuild for 1.7.17 that has some more features: https://github.com/devurandom/gentoo-overlay/blob/master/media-libs/openimageio/openimageio-1.7.17.ebuild

In particular, the PYTHON_REQUIRED_USE is only required if USE=python is set and I can select between jpeg2k and jpegturbo. My media-libs/field3d has a subslot, so I can `:=` depend on it. I also depend on `:0=` of media-libs/giflib. If USE=python, I also depend on the media-libs/opencv[python,$PYTHON_USEDEP], and also require a more recent version. I also consider CHANGES.md, CREDITS.md and README.md as documentation and I run `python_fix_shebang`, only if USE=python (might be possible to also do this otherwise, I don't recall). I have a slightly different way of assembling -DUSE_SIMD, which might be interesting to you. I explicitly disable static libraries and I also disable Nuke support, since Gentoo has no such package. I enable a verbose build. I set -DUSE_PYTHON/3 depending on USE=python and which Python interpreter we are building for.

There seems to be a bug in this ebuild, where you depend twice on virtual/jpeg, once unconditionally and once behind `jpeg?`.

devurandom commented:
P.S. [My ebuild](https://github.com/devurandom/gentoo-overlay/blob/master/media-libs/openimageio/openimageio-1.7.17.ebuild) is based on my suggestion on the Gentoo bugtracker, probably with one or the other refinement: https://bugs.gentoo.org/596268#c3

dracwyrm commented:
@devurandom setting -DUSE_PYTHON is no longer needed with my patch set as boost and python are completely fixed. I ran multiple tests on it.

I did have that OpenCV depend you have but it completely failed the check not. I'll retry it.

I'll have a look later at your ebuild.

dracwyrm commented:
@devurandom I just pushed the latest updates.

I have been debating a long time -- as in months -- about leaving verbose on in production ebuilds. I have it on for testing, but then remove it for production. I'm still debating.
 
I removed all python stuff from my local ebuild when python didn't work, and then put it back in, so I forgot the required deps python test. My bad. Thanks for pointing that.

The virtual jpeg was in there twice because the original ebuild hard depended on it and I added a use flag later after going through the source code and found it to be optional. I just forgot the hard depend on one. virtual jpeg is jpeg turbo, so I left that out, but added in jpeg2k. My original idea was to add that in but mis-typed it as just jpeg.

python_fix_shebang isn't needed with these files which is why it was removed. ;) It's not an always needed function and only for when upstream does some crazy things to their python shebangs.

I improved on your SIMD code. :D 

Thanks for your suggestions and pointing out some of my oopsies.

devurandom commented:
> setting -DUSE_PYTHON is no longer needed with my patch set as boost and python are completely fixed. I ran multiple tests on it.

How do you respect USE=python? I.e. the Gentoo variable?

> leaving verbose on in production ebuilds

I am all in favour. It makes debugging of end-user build issues a lot easier. I argued the same way for dev-lang/rust.

> Thanks for your suggestions and pointing out some of my oopsies.

I'm glad to have been of service.

devurandom commented:
Regarding jpeg-turbo: I was not sure whether upstream actually supports non-libjpeg-turbo (e.g. the old libjpeg.so provided by virtual/jpeg). Hence I decided to not depend on virtual/jpeg, but media-libs/libjpeg-turbo explicitly. I assume you confirmed that upstream officially supports non-libjpeg-turbo, but just named the CMake flag in a weird way?

Does Python support in oiio depend on Python support in OpenCV? That's what I apparently figured out back in the day -- at least when I look at the dependency string that I constructed to mirror upstream requirements.

Rest seems fine now. I like the `join_by` function more than inlining the code, but that's more or less preference / readability (assuming not everyone knows from the top of their head about the weird interaction between `$*` or `${array[*]}` and `$IFS` in Bash).

dracwyrm commented:
> How do you respect USE=python? I.e. the Gentoo variable?

I meant USE_PYTHON3 and co. USE_PYTHON is now for everything. Sorry. Was a long day when I typed that, and having a cold doesn't make it easier. :D So, setting USE_PYTHON enables python correctly for all versions installed, even if multiple are installed. I have 2.7, 3.4, and 3.5 and was able to select each one correctly. :) I'm currently working on code to see if I can compile and install the module for each installed version of python, not just one. 

> I am all in favour. It makes debugging of end-user build issues a lot easier. I argued the same way for dev-lang/rust.

Those are the points I'm thinking of. Though, the trouble with OIIO, is that a lot of the message outputs are behind custom variables. Actual code example:
```
if (NOT OpenEXR_FIND_QUIETLY)
    message (STATUS "OPENEXR_INCLUDE_DIR = ${OPENEXR_INCLUDE_DIR}")
    message (STATUS "OPENEXR_LIBRARIES = ${OPENEXR_LIBRARIES}")
endif ()
```
There are variables like that for each package that OIIO depends on, so in this case, verbose will not display everything. What I'm working on at the moment is a patch for upstream that puts all status messages under verbose and see if they except it for RB-1.7, RB-1.8, and master. Otherwise status messages will be a bit broken.

> Regarding jpeg-turbo: I was not sure whether upstream actually supports non-libjpeg-turbo (e.g. the old libjpeg.so provided by virtual/jpeg). Hence I decided to not depend on virtual/jpeg, but media-libs/libjpeg-turbo explicitly. I assume you confirmed that upstream officially supports non-libjpeg-turbo, but just named the CMake flag in a weird way?

I just had another look at the code as it's been ages since I did that, since way before I posted the original 1.7.7 ebuild on b.g.o. The problem is this bit:
```
if (USE_JPEGTURBO)
    find_package (JPEGTurbo)
endif ()
if (NOT JPEG_FOUND)
    find_package (JPEG REQUIRED)
endif ()
```
It hard depends on some form of JPEG being installed, which is why the virtual/jpeg was not under a useflag. OIIO will choose JPEG Turbo over the normal one, but there isn't a global use flag for jpeg turbo, which means a custom one. So new code without a virtual:
```
jpegturbo? ( media-libs/libjpeg-turbo )
!jpegturbo? (media-libs/jpeg )
```

> Does Python support in oiio depend on Python support in OpenCV?

I looked through all the files in scr/python and found no mention of references to opencv at all, so it might not need the python bindings of opencv.

devurandom commented:
Sorry, I overlooked `USE_PYTHON` in `mycmakeargs`, which triggers on `USE=python`.

Verbosity: Thanks for the work!

jpeg vs. jpegturbo: Does oiio actually use jpeg-turbo specific APIs? A quick ripgrep through the code does not show any reference to jpeg-turbo except in the build-system. Do we know why the specifically check for `JPEGTurbo`? Why they are not just checking for `JPEG` only or why they hide jpeg-turbo behind `USE_JPEGTURBO`? Maybe upstream's build-system code regarding detection of the libjpeg implementation could be simplified, and we could use virtual/jpeg without additional use-flags afterwards?

Python + OpenCV: No idea where I got the different version dependencies and use-flag dependencies from, but I see you had a much deeper look into the code, than I probably had.

dracwyrm commented:
I think the APIs for normal jpeg and jpeg turbo are the same, as turbo is meant to be a drop in replacement. The only difference is the backend code that call specific SIMDs of the processor to make processing jpegs faster. That's the way I understand it. So, I'm leaning more towards just depending on virtual/jpeg like the original ebuild and letting the user pick from there, though they most likely have jpeg installed.

monsieurp commented on media-libs/openimageio/openimageio-1.7.17.ebuild:
```diff
@@ -0,0 +1,112 @@
+	[[ -z ${mysimd} ]] && mysimd=("0")
+
+	local mycmakeargs=(
+		-DCMAKE_INSTALL_DOCDIR="share/doc/${PF}"
```
It doesn't look right, does it? isn't `/usr/share/doc/${PF}` the correct path? Ditto for the other ebuild.

SoapGentoo replied:
it is absolutely correct. If the path given to `CMAKE_INSTALL_DOCDIR` is relative (which it is), cmake automatically prepends `CMAKE_INSTALL_PREFIX` (which is `${EPREFIX}/usr`).

a17r replied:
@monsieurp ping

SoapGentoo commented on media-libs/openimageio/openimageio-1.6.18.ebuild:
```diff
@@ -0,0 +1,108 @@
+	>=media-libs/libwebp-0.2.1:=
+	media-libs/openexr:=
+	media-libs/tiff:0=
+	sys-libs/zlib
```
`:=`

SoapGentoo commented on media-libs/openimageio/openimageio-1.6.18.ebuild:
```diff
@@ -0,0 +1,108 @@
+	truetype? ( media-libs/freetype:2= )"
+DEPEND="${RDEPEND}"
+
+PATCHES=( "${FILESDIR}/${P}-use-gnuinstalldirs.patch"
```
put this on the next line

SoapGentoo commented on profiles/base/package.use.mask:
```diff
@@ -275,9 +275,11 @@ sys-devel/gcc vtv
 # invokes it wrong.
 dev-lang/moarvm asan
 
+# Jonathan Scruggs <j.scruggs@gmail.com> (28/09/2017)
```
1. use the proper Gentoo date format
2. chuck out kensington's author line

dracwyrm replied:
Okay. Sorry. I thought we had to keep credit to the first person that placed it there.

monsieurp replied:
Sorry for coming across as annoying if you want to keep the credit just use this format:
```
Some Gentoo Dev Who Will Merge This PR <someone@gentoo.org>
on behalf of Jonathan Scruggs <j.scruggs@gmail.com> (28 Sep 2017)
```

We do this to keep track of what's going on with these entries. Allowing "external" domain names in these files will confuse the hell out of developers and people after a while.

monsieurp replied:
Could you please update it accordingly? Thanks.

monsieurp replied:
Of course "Some Gentoo Dev" is an example. Use @kensington or me if you want. We will modify it on the fly before merging.

dracwyrm replied:
@monsieurp @SoapGentoo I added that info like you asked. Sorry. Been really busy with work and knackered. :(

dracwyrm commented:
@monsieurp @SoapGentoo Made requested changes. Are there anymore that need to be done? Hopefully compile test goes alright. OpenImageIO 1.7 is a big departure from 1.6, which is why I updated 1.6 branch to be made the new stable in several weeks time ... hopefully.

devurandom commented:
What's the status of this?

dracwyrm commented:
I was wondering this myself as I can't push a Blender PR until this is merged.
